File: sdk/ruby/lib/xid/session_token_exchange.rb
# frozen_string_literal: true

module Xid
  SessionTokenHttpResponse = Struct.new(:status, :body, keyword_init: true)

  module SessionTokenExchange
    SESSION_TOKEN_PATH = "/v1/sessions/token"

    module_function

    def exchange(incoming_request_url:, cookie_header:, endpoint: nil, transport: nil)
      resolved = resolve_endpoint(incoming_request_url, endpoint)
      response = if transport
                   transport.call(resolved, cookie_header)
                 else
                   post_without_redirect(resolved, cookie_header)
                 end
      status = Integer(response.status)
      unless status == 200
        raise SessionTokenExchangeError, "Session token exchange returned HTTP #{status}"
      end

      body = JSON.parse(response.body)
      unless body.is_a?(Hash) &&
             body.keys == ["token"] &&
             body["token"].is_a?(String) &&
             !body["token"].strip.empty?
        raise SessionTokenExchangeError, "Session token exchange returned an invalid response"
      end
      body["token"]
    rescue SessionTokenExchangeError
      raise
    rescue JSON::ParserError => e
      raise SessionTokenExchangeError, "Session token exchange returned invalid JSON: #{e.message}"
    rescue StandardError => e
      raise SessionTokenExchangeError, "Session token exchange failed: #{e.message}"
    end

    def resolve_endpoint(incoming_request_url, endpoint)
      incoming = URI.parse(incoming_request_url)
      unless http_uri?(incoming) && incoming.userinfo.nil?
        raise SessionTokenExchangeError,
              "Incoming request URL must be an absolute HTTP(S) URL"
      end
      resolved = URI.join(incoming.to_s, endpoint || SESSION_TOKEN_PATH)
      unless http_uri?(resolved) &&
             resolved.userinfo.nil? &&
             origin(incoming) == origin(resolved) &&
             resolved.path == SESSION_TOKEN_PATH &&
             resolved.query.nil? &&
             resolved.fragment.nil?
        raise SessionTokenExchangeError,
              "Session token endpoint must be exact same-origin #{SESSION_TOKEN_PATH}"
      end
      resolved
    rescue URI::InvalidURIError => e
      raise SessionTokenExchangeError, "Invalid session token exchange URL: #{e.message}"
    end

    def post_without_redirect(uri, cookie_header)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.is_a?(URI::HTTPS)
      http.open_timeout = 10
      http.read_timeout = 10
      request = Net::HTTP::Post.new(uri.request_uri)
      request["Accept"] = "application/json"
      request["Cookie"] = cookie_header
      response = http.request(request)
      SessionTokenHttpResponse.new(status: response.code.to_i, body: response.body.to_s)
    end

    def http_uri?(uri)
      uri.is_a?(URI::HTTP) && uri.absolute? && !uri.host.to_s.empty?
    end

    def origin(uri)
      "#{uri.scheme.downcase}://#{uri.host.downcase}:#{uri.port}"
    end
  end
end

File: packages/backend/src/session-token-exchange.ts
import type { Result, SessionTokenResponse } from '@xid-kit/types'

import { AppError } from './errors'

const CORE_REFRESH_COOKIE_PREFIX = '__Host-xid.rt.'
const DEFAULT_SESSION_TOKEN_ENDPOINT = '/v1/sessions/token'

export type SessionTokenExchangeError = 'no_core_session' | 'session_rejected'

export type SessionTokenExchangeOptions = {
  // 相对路径按 request 解析,解析后必须仍在 exact same origin。
  endpoint?: string
  // 须服务端 fetch:会转发入站 Request 的 HttpOnly Cookie,禁止浏览器 fetch。
  fetcher?: typeof fetch
  signal?: AbortSignal
}

function cookieNames(request: Request): string[] {
  const header = request.headers.get('cookie')
  if (!header) return []

  return header.split(';').flatMap((part) => {
    const separator = part.indexOf('=')
    if (separator === -1) return []
    const name = part.slice(0, separator).trim()
    return name ? [name] : []
  })
}

export function hasCoreSessionCookie(request: Request): boolean {
  return cookieNames(request).some((name) => name.startsWith(CORE_REFRESH_COOKIE_PREFIX))
}

function resolveSameOriginEndpoint(request: Request, endpoint?: string): URL {
  const requestUrl = new URL(request.url)
  const resolved = new URL(endpoint ?? DEFAULT_SESSION_TOKEN_ENDPOINT, requestUrl)

  if (
    resolved.origin !== requestUrl.origin ||
    resolved.username ||
    resolved.password ||
    resolved.pathname !== DEFAULT_SESSION_TOKEN_ENDPOINT ||
    resolved.search ||
    resolved.hash
  ) {
    throw new AppError(
      'invalid_options',
      'sessionTokenExchange.endpoint must be exact same-origin /v1/sessions/token',
    )
  }

  return resolved
}

function parseSessionTokenResponse(value: unknown): SessionTokenResponse | null {
  if (value === null || typeof value !== 'object') return null
  if (Array.isArray(value) || Object.keys(value).length !== 1) return null
  const token = (value as { token?: unknown }).token
  if (typeof token !== 'string' || token.trim().length === 0) return null
  return { token }
}

// opaque refresh 不本地解析;整段 Cookie 仅转发 exact same-origin,因多会话选择依赖 HttpOnly active-session 指针。
export async function exchangeSessionToken(
  request: Request,
  options: SessionTokenExchangeOptions = {},
): Promise<Result<SessionTokenResponse, SessionTokenExchangeError>> {
  if (!hasCoreSessionCookie(request)) {
    return { ok: false, error: 'no_core_session' }
  }

  const cookie = request.headers.get('cookie')
  if (!cookie) {
    return { ok: false, error: 'no_core_session' }
  }

  const endpoint = resolveSameOriginEndpoint(request, options.endpoint)
  const fetcher = options.fetcher ?? globalThis.fetch.bind(globalThis)

  let response: Response
  try {
    response = await fetcher(endpoint, {
      method: 'POST',
      headers: {
        accept: 'application/json',
        cookie,
      },
      redirect: 'manual',
      ...(options.signal ? { signal: options.signal } : {}),
    })
  } catch (cause) {
    throw new AppError(
      'session_token_exchange_failed',
      'Core session-token exchange request failed',
      { cause },
    )
  }

  if (response.status === 401 || response.status === 403) {
    return { ok: false, error: 'session_rejected' }
  }
  if (!response.ok) {
    throw new AppError(
      'session_token_exchange_failed',
      `Core session-token exchange returned HTTP ${response.status}`,
    )
  }

  let body: unknown
  try {
    body = await response.json()
  } catch (cause) {
    throw new AppError(
      'session_token_exchange_failed',
      'Core session-token exchange returned invalid JSON',
      { cause },
    )
  }

  const parsed = parseSessionTokenResponse(body)
  if (!parsed) {
    throw new AppError(
      'session_token_exchange_failed',
      'Core session-token exchange returned an invalid response',
    )
  }
  return { ok: true, value: parsed }
}
